File: NewsDaily/app/src/main/java/com/andruid/magic/newsdaily/activity/SearchActivity.java
package com.andruid.magic.newsdaily.activity;

import android.os.Bundle;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.view.animation.AccelerateInterpolator;

import androidx.annotation.NonNull;
import androidx.appcompat.app.ActionBar;
import androidx.appcompat.app.AppCompatActivity;
import androidx.databinding.DataBindingUtil;
import androidx.lifecycle.ViewModelProvider;

import com.andruid.magic.newsdaily.R;
import com.andruid.magic.newsdaily.adapter.NewsAdapter;
import com.andruid.magic.newsdaily.articles.ArticlesViewModel;
import com.andruid.magic.newsdaily.articles.ArticlesViewModelFactory;
import com.andruid.magic.newsdaily.databinding.ActivitySearchBinding;
import com.andruid.magic.newsdaily.util.RxSearchObservable;
import com.yuyakaido.android.cardstackview.CardStackLayoutManager;
import com.yuyakaido.android.cardstackview.CardStackListener;
import com.yuyakaido.android.cardstackview.Direction;
import com.yuyakaido.android.cardstackview.Duration;
import com.yuyakaido.android.cardstackview.StackFrom;
import com.yuyakaido.android.cardstackview.SwipeAnimationSetting;

import java.util.concurrent.TimeUnit;

import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.Disposable;
import io.reactivex.schedulers.Schedulers;
import timber.log.Timber;

public class SearchActivity extends AppCompatActivity {
    private ActivitySearchBinding binding;
    private Disposable disposable;
    private NewsAdapter newsAdapter;
    private ArticlesViewModel articlesViewModel;
    private CardStackLayoutManager cardStackLayoutManager;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        binding = DataBindingUtil.setContentView(this, R.layout.activity_search);
        setSupportActionBar(binding.toolBar);
        ActionBar actionBar = getSupportActionBar();
        if(actionBar != null)
            actionBar.setDisplayHomeAsUpEnabled(true);
        newsAdapter = new NewsAdapter();
        cardStackLayoutManager = new CardStackLayoutManager(this, new CardStackListener() {
            @Override
            public void onCardDragging(Direction direction, float ratio) {}
            @Override
            public void onCardSwiped(Direction direction) {}
            @Override
            public void onCardRewound() {}
            @Override
            public void onCardCanceled() {}
            @Override
            public void onCardAppeared(View view, int position) {}
            @Override
            public void onCardDisappeared(View view, int position) {}
        });
        articlesViewModel = new ViewModelProvider(this,
                new ArticlesViewModelFactory("en")).get(ArticlesViewModel.class);
        setUpCardStackView();
        disposable = RxSearchObservable.fromView(binding.searchView)
                .debounce(1000, TimeUnit.MILLISECONDS)
                .filter(text -> !text.isEmpty())
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(this::loadArticles);
    }

    private void setUpCardStackView() {
        SwipeAnimationSetting swipeSetting = new SwipeAnimationSetting.Builder()
                .setDirection(Direction.Bottom)
                .setInterpolator(new AccelerateInterpolator())
                .setDuration(Duration.Normal.duration)
                .build();
        cardStackLayoutManager.setSwipeAnimationSetting(swipeSetting);
        cardStackLayoutManager.setCanScrollHorizontal(false);
        cardStackLayoutManager.setDirections(Direction.VERTICAL);
        cardStackLayoutManager.setStackFrom(StackFrom.Bottom);
        binding.cardStackView.setLayoutManager(cardStackLayoutManager);
        binding.cardStackView.setAdapter(newsAdapter);
    }

    private void loadArticles(String query) {
        articlesViewModel.loadArticles(query).observe(this, news -> {
            Timber.d("search for %s : %d results", query, news.size());
            newsAdapter.submitList(news);
            setTitle("Search results for "+query);
        });
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.menu_search, menu);
        MenuItem item = menu.findItem(R.id.searchItem);
        binding.searchView.setMenuItem(item);
        binding.searchView.showSearch(true);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(@NonNull MenuItem item) {
        if (item.getItemId() == android.R.id.home)
            onBackPressed();
        return true;
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        binding.unbind();
        disposable.dispose();
    }

    @Override
    public void onBackPressed() {
        if (binding.searchView.isSearchOpen()) {
            binding.searchView.closeSearch();
        } else {
            super.onBackPressed();
        }
    }
}

File: NewsDailyJava/app/src/main/java/com/andruid/magic/newsdaily/activity/SearchActivity.kt
package com.andruid.magic.newsdaily.activity

import android.os.Bundle
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.view.animation.AccelerateInterpolator
import androidx.appcompat.app.AppCompatActivity
import androidx.databinding.DataBindingUtil
import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModelProvider
import com.andruid.magic.newsdaily.R
import com.andruid.magic.newsdaily.adapter.NewsAdapter
import com.andruid.magic.newsdaily.articles.ArticlesViewModel
import com.andruid.magic.newsdaily.databinding.ActivitySearchBinding
import com.andruid.magic.newsdaily.util.BaseViewModelFactory
import com.andruid.magic.newsdaily.util.RxSearchObservable.Companion.fromView
import com.yuyakaido.android.cardstackview.*
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.Disposable
import io.reactivex.schedulers.Schedulers
import java.util.concurrent.TimeUnit

class SearchActivity : AppCompatActivity() {
    private lateinit var binding: ActivitySearchBinding
    private lateinit var disposable: Disposable
    private lateinit var articlesViewModel: ArticlesViewModel
    private lateinit var cardStackLayoutManager: CardStackLayoutManager

    private var newsAdapter: NewsAdapter = NewsAdapter()

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = DataBindingUtil.setContentView(this, R.layout.activity_search)
        cardStackLayoutManager = CardStackLayoutManager(this, object : CardStackListener {
            override fun onCardDragging(direction: Direction, ratio: Float) {}
            override fun onCardSwiped(direction: Direction) {}
            override fun onCardRewound() {}
            override fun onCardCanceled() {}
            override fun onCardAppeared(view: View, position: Int) {}
            override fun onCardDisappeared(view: View, position: Int) {}
        })
        articlesViewModel = ViewModelProvider(this, BaseViewModelFactory {
            ArticlesViewModel("en") }).get(ArticlesViewModel::class.java)
        disposable = fromView(binding.searchView)
                .debounce(1000, TimeUnit.MILLISECONDS)
                .filter { text: String -> text.isNotEmpty() }
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe { query: String? -> loadArticles(query ?: "") }

        setSupportActionBar(binding.toolBar)
        supportActionBar?.setDisplayHomeAsUpEnabled(true)

        setUpCardStackView()

        articlesViewModel.pagedListLiveData.observe(this, Observer {
            newsAdapter.submitList(it)
        })
    }

    private fun setUpCardStackView() {
        cardStackLayoutManager.apply {
            val swipeSetting = SwipeAnimationSetting.Builder()
                    .setDirection(Direction.Bottom)
                    .setInterpolator(AccelerateInterpolator())
                    .setDuration(Duration.Normal.duration)
                    .build()
            setSwipeAnimationSetting(swipeSetting)
            setCanScrollHorizontal(false)
            setDirections(Direction.VERTICAL)
            setStackFrom(StackFrom.Bottom)
        }
        binding.cardStackView.apply {
            layoutManager = cardStackLayoutManager
            adapter = newsAdapter
        }
    }

    fun loadArticles(query : String) {
        articlesViewModel.setQuery(query)
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.menu_search, menu)
        val item = menu.findItem(R.id.searchItem)
        binding.searchView.setMenuItem(item)
        binding.searchView.showSearch(true)
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        if (item.itemId == android.R.id.home)
            onBackPressed()
        return true
    }

    override fun onDestroy() {
        super.onDestroy()
        binding.unbind()
        disposable.dispose()
    }

    override fun onBackPressed() {
        if (binding.searchView.isSearchOpen)
            binding.searchView.closeSearch()
        else
            super.onBackPressed()
    }
}